### PR TITLE
User interface

### PR DESCRIPTION
**Pull request by ImFriedHof**
This marks the merge of the UserInterfaceBranch to the main branch. 

In this branch the UI was created and it has the following structure: 

1. A frame that holds both the GraphPanel and the ControlPanel. (The ratio between these two is managed by a GridBagLayout with weighting )

2. The GraphPanel is taken from Mr. Noll's Code which was used as a codebase for this project. The ControlPanel itself is a TabbedPane which is used to select between the parametric and the polynomial functions. 
The two Panels that are created by this TabbedPane are named PolynomialFunctionControl and ParametricFunctionControl. 

3. PolynomialFunctionControl: This Panel (in combination with the FunctionControlPanel) is the most complicated because it stacks Panel into Panel because it is the only way to set a working Layout. For visual help I added a screenshot of a sketch that I made.
Bild_2023-06-30_172558174

The AddButton in the PolynomialHelpPanel adds one instance of the FunctionControlPanel. This is currently capped by 3, but could be increased to a max of about 7. 
The calculateButton gets the Text of the InputField and returns the ID/count of the instance it came from.
The deleteButton deletes its Instance of FunctionControlPanel.

4. The ParametricFunctionControl: This panel faces the same complexity issues as the Polynomial ome but in smaller. Its purpose is to be a layout panel to the ParametricHelpPanel and the ParametricFunctionControlPanel which are glued to the bottom and center respectively. 
The ParametricHelpPanel is a smaller version of the PoynomialHelpPanel (it misses the AddButton). 

The ParametricFunctionControlPanel implements the actual function input. It holds two FunctionFields for the X and Y function, two boxes for setting the range (which are proteted against nonNumeric inputs exept null and ".")  and a CalculateButton and a DeleteButton. 

The CalculateButton gets the input from the user in the textFields and the DeleteButton deletes all user Input.

An idea of mine is to rename some of those classes for more intuitive understanding of its contents as well as to further define the differenes to other classes and their names. 

FunctionControlPanel --> PolynomialFunctionInput
ParametricFunctionControlPanel --> ParametricFunctionInput

### CODE REVIEW

File: .idea/misc.xml
Comment (kantholz44mm) [inline .idea/misc.xml:1]: The .idea folder does not belong on the Git repo. It's currently included in the .gitignore file, please check if you got these changes.

File: .idea/modules.xml
Comment (kantholz44mm) [inline .idea/modules.xml:1]: Neither does this

File: .idea/vcs.xml
Comment (kantholz44mm) [inline .idea/vcs.xml:1]: Neither does this

File: PolynomialPlotter.iml
Comment (kantholz44mm) [inline PolynomialPlotter.iml:1]: Neither does this. Please add this to the .gitignore file.

File: src/ControlPanel.java
Comment (kantholz44mm) [inline src/ControlPanel.java]: Missing explicit visibility (public, private, protected, etc.)

File: src/ControlPanel.java
Comment (kantholz44mm) [inline src/ControlPanel.java]: Fine for testing, but this should be changed before merging

File: src/Frame.java
Comment (kantholz44mm) [inline src/Frame.java]: Missing explicit visibility as well

File: src/Frame.java
Comment (kantholz44mm) [inline src/Frame.java]: The name 'Frame' is a bit redundant as is. Maybe change to something like 'GraphPlotterFrame' etc.

File: src/FunctionControlPanel.java
Comment (kantholz44mm) [inline src/FunctionControlPanel.java]: Ambiguous variable name. What colors? Maybe change to 'GraphColors' etc. Also, inconsistent naming, starting with capital letters for some and regular ones for most other variables. Also missing explicit visibility specifier

File: src/FunctionControlPanel.java
Comment (kantholz44mm) [inline src/FunctionControlPanel.java]: count of what?

File: src/FunctionControlPanel.java
Comment (kantholz44mm) [inline src/FunctionControlPanel.java]: Same as red background color for control panel, should be changed before merging

File: src/FunctionControlPanel.java
Comment (kantholz44mm) [inline src/FunctionControlPanel.java]: Maybe we can make use of JColorChooser:
https://docs.oracle.com/javase/tutorial/uiswing/components/colorchooser.html

(Optional, nice to have, but not required.)

File: src/ParametricFunctionControl.java
Comment (kantholz44mm) [inline src/ParametricFunctionControl.java]: missing explicit visibility specifier

File: src/ParametricFunctionControlPanel.java
Comment (kantholz44mm) [inline src/ParametricFunctionControlPanel.java]: all of these are missing explicit visibility specifiers

File: src/ParametricFunctionControlPanel.java
Comment (kantholz44mm) [inline src/ParametricFunctionControlPanel.java:1]: does the input field for ranges also accept negative numbers?

File: src/PolynomialFunctionControl.java
Comment (kantholz44mm) [inline src/PolynomialFunctionControl.java]: missing visibility specifier

File: src/PolynomialFunctionControl.java
Comment (kantholz44mm) [inline src/PolynomialFunctionControl.java]: please remove debug outputs before merging

File: src/PolynomialFunctionControl.java
Comment (kantholz44mm) [inline src/PolynomialFunctionControl.java]: missing explicit visibility specifier

File: src/PolynomialFunctionControl.java
Comment (kantholz44mm) [inline src/PolynomialFunctionControl.java]: missing explicit visibility specifier

**Review by kantholz44mm — CHANGES_REQUESTED** [review]
Works in general, but there are some things that should be fixed before merging

File: src/FunctionControlPanel.java
Comment (FabianRolfMatthiasNoll) [reply src/FunctionControlPanel.java]: As far as i know he tried it but it made things complicated. Would put this maybe on hold for a future "feature"?

File: src/ParametricFunctionControlPanel.java
Comment (FabianRolfMatthiasNoll) [inline src/ParametricFunctionControlPanel.java]: Is this comment still needed? Uncommented Code is generally not a good practice. Also please review all your comments if they are really needed and useful

File: src/ParametricFunctionControlPanel.java
Comment (FabianRolfMatthiasNoll) [inline src/ParametricFunctionControlPanel.java:190]: Regarding this function. You do not need to assign the result of parseDouble to a variable if you are not going to use it.
Also in the first if statment the string could be empty (.isEmpty exists) maybe it would be useful to check for that also. Also move the return true in the try block

File: src/Frame.java
Comment (FabianRolfMatthiasNoll) [inline src/Frame.java]: Please review all the comments if they are useful. Many times the code is explaining itself more than enough. In that case pls remove the comments

File: src/FunctionControlPanel.java
Comment (FabianRolfMatthiasNoll) [inline src/FunctionControlPanel.java]: Why are the other ActionListener private and this one public?

File: src/ParametricFunctionControl.java
Comment (FabianRolfMatthiasNoll) [inline src/ParametricFunctionControl.java:18]: Shouldnt the HelpTabel the exact same for both Panels? This would result in a double implementation and double change if changes come up. Make it a separate class and use it in both Tabs if it has to be tab specific

File: src/PolynomialFunctionControl.java
Comment (FabianRolfMatthiasNoll) [inline src/PolynomialFunctionControl.java:45]: Same thing as in the ParametricFunctionControl ^^

File: src/ParametricFunctionControl.java
Comment (ImFriedHof) [reply src/ParametricFunctionControl.java:18]: The problem is that the AddButton is not needed in the parametricFunction tab as there is only one function at a time. But it is necessary in the polynomialFunction tab.
You are right, it is a double implementation of the 3 other buttons, but the uses of the Buttons also change. For example the DeleteButton in the polynomialFunction tab is used for deleting an entire instance of the FunctionControlPanel.
In the parametricFunction tab it is used for removing all the users inputs out of the TextFields but it does not delete the TextFields itself.
There sure is a better way to implement this but I dont know how to do it.

File: .idea/misc.xml
Comment (ImFriedHof) [reply .idea/misc.xml:1]: Used the answer you gave me. Please check if it worked correctly. From what I can tell its better now.
This of course applies to the other files as well. I will not comment under all of them. This one should suffice.

File: src/ControlPanel.java
Comment (ImFriedHof) [reply src/ControlPanel.java]: Done

File: src/ControlPanel.java
Comment (ImFriedHof) [reply src/ControlPanel.java]: Done

File: src/Frame.java
Comment (ImFriedHof) [reply src/Frame.java]: Done

File: src/Frame.java
Comment (ImFriedHof) [reply src/Frame.java]: Done

File: src/FunctionControlPanel.java
Comment (ImFriedHof) [reply src/FunctionControlPanel.java]: Done

File: src/FunctionControlPanel.java
Comment (ImFriedHof) [reply src/FunctionControlPanel.java]: Done

File: src/ParametricFunctionControl.java
Comment (ImFriedHof) [reply src/ParametricFunctionControl.java]: Done

File: src/ParametricFunctionControlPanel.java
Comment (ImFriedHof) [reply src/ParametricFunctionControlPanel.java]: Done

File: src/PolynomialFunctionControl.java
Comment (ImFriedHof) [reply src/PolynomialFunctionControl.java]: Done

File: src/PolynomialFunctionControl.java
Comment (ImFriedHof) [reply src/PolynomialFunctionControl.java]: Done

File: src/PolynomialFunctionControl.java
Comment (ImFriedHof) [reply src/PolynomialFunctionControl.java]: Done

File: src/PolynomialFunctionControl.java
Comment (ImFriedHof) [reply src/PolynomialFunctionControl.java]: Done

File: src/ParametricFunctionControlPanel.java
Comment (ImFriedHof) [reply src/ParametricFunctionControlPanel.java]: Done

File: src/Frame.java
Comment (ImFriedHof) [reply src/Frame.java]: Done

File: src/FunctionControlPanel.java
Comment (ImFriedHof) [reply src/FunctionControlPanel.java]: Changed

File: src/ParametricFunctionControlPanel.java
Comment (ImFriedHof) [reply src/ParametricFunctionControlPanel.java:1]: It does now.

File: src/ParametricFunctionControlPanel.java
Comment (ImFriedHof) [reply src/ParametricFunctionControlPanel.java:190]: I removed the variable and moved the return statement.
But I do not understand the need to check for an empty String. This function only checks user input and the user can not input "empty". The only thing the user can do is delete the complete String written in it, which is handled by the replace function.

Comment (FabianRolfMatthiasNoll) [reply src/ParametricFunctionControlPanel.java:190]: I didnt look completely at the code. What happens when a user inserts a empty string? Because if he does it is not null it just has the length of 0 so "". Does it get filterd beforehand?

File: src/FunctionControlPanel.java
Comment (ImFriedHof) [reply src/FunctionControlPanel.java]: Changed to CountOfInstances

File: src/ParametricFunctionControl.java
Comment (FabianRolfMatthiasNoll) [reply src/ParametricFunctionControl.java:18]: Ok then we will leave it for now. But maybe (we still have time) can split it into two panels. A general Panel (Screenshot, Help etc.) and a specific function control panel (polynomial: Cam Reset, Add? / parametric etc)

File: .idea/misc.xml
Comment (FabianRolfMatthiasNoll) [reply .idea/misc.xml:1]: The .iml is still there

File: PolynomialPlotter.iml
Comment (FabianRolfMatthiasNoll) [reply PolynomialPlotter.iml:1]: Still in the Repo. Not solved

File: src/FunctionControlPanel.java
Comment (ImFriedHof) [reply src/FunctionControlPanel.java]: Done

File: src/ParametricFunctionControlPanel.java
Comment (FabianRolfMatthiasNoll) [reply src/ParametricFunctionControlPanel.java:190]: Done

File: src/ParametricFunctionControl.java
Comment (FabianRolfMatthiasNoll) [reply src/ParametricFunctionControl.java:18]: Done

**Review by kantholz44mm — APPROVED** [review]

**Pull request merged**
